### PR TITLE
[M5] Mina libp2p network basic setup

### PR DESCRIPTION
**Pull request by hanabi1224**
**Summary of changes**
Changes introduced in this pull request:
- Mina libp2p network basic setup

**Reference issue to close (if applicable)**

Closes 

**Other information and links**

### CODE REVIEW

File: network/src/p2p/builder.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+use std::{borrow::Borrow, time::Duration};
+
+#[derive(Clone)]
+pub struct MinaTransportBuilder {
+    pnet_config: PnetConfig,
+    mplex_config: MplexConfig,
+    noise_keys: AuthenticKeypair<X25519Spec>,
+    timeout: Duration,
+    peer_id: PeerId,
+}
```
Comment (polarbeargo) [inline network/src/p2p/builder.rs:23]: May you share a reference where this struct has been defined?

File: network/src/p2p/builder.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            self.peer_id,
+        )
+    }
+}
+
+impl Default for MinaTransportBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
```
Comment (polarbeargo) [inline network/src/p2p/builder.rs:100]: Excellent work on defining and impl MinaTransportBuilder 👍 !  May you share a reference where this struct has been defined?

File: network/src/p2p/builder.rs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        keypair: identity::Keypair,
+        config: &MinaTransportConfig,
+    ) -> Self {
+        let peer_id = PeerId::from(keypair.public());
+        let shared_key = config.get_shared_key();
+        let pnet_config = PnetConfig::new(shared_key);
+        let mplex_config = config.get_mplex_config();
+        let noise_keys = noise::Keypair::<noise::X25519Spec>::new()
+            .into_authentic(&keypair)
+            .expect("Signing libp2p-noise static DH keypair failed.");
```
Comment (polarbeargo) [inline network/src/p2p/builder.rs:45]: Good work on use `.expect()`

File: network/src/p2p/config.rs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        let mut hasher = Blake2b256::default();
+        hasher.update(self.rendezvous_string);
+        let hash = hasher.finalize();
+        let mut psk_fixed: [u8; 32] = Default::default();
+        psk_fixed.copy_from_slice(hash.as_ref());
+        PreSharedKey::new(psk_fixed)
+    }
+
+    pub fn get_mplex_config(&self) -> MplexConfig {
+        let mut config = MplexConfig::new();
```
Comment (polarbeargo) [inline network/src/p2p/config.rs:33]: Excellent work on implementing a new method for MinaTransportBuilder much cleaner, nit 👍 !

File: network/src/p2p/builder.rs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        Self {
+            pnet_config,
+            mplex_config,
+            noise_keys,
+            timeout: Duration::from_secs(60),
+            peer_id,
```
Comment (polarbeargo) [inline network/src/p2p/builder.rs:51]: Does timeout already define to 60 sec?

File: network/src/p2p/config.rs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+#[derive(Clone, Debug)]
+pub struct MinaTransportConfig<'a> {
+    pub rendezvous_string: &'a [u8],
+    pub mplex_protocol_name: &'static [u8],
+}
+
+impl<'a> MinaTransportConfig<'a> {
+    pub fn get_shared_key(&self) -> PreSharedKey {
+        let mut hasher = Blake2b256::default();
```
Comment (polarbeargo) [inline network/src/p2p/config.rs:24]: Good work integrating  `Blake2b256` struct in crate `libp2p

File: network/tests/connection_mainnet.rs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+                let builder = MinaTransportBuilder::default()
+                    .with_config(config)
+                    .with_timeout(Duration::from_secs(timeout_secs));
+                let tcp = TcpConfig::new().nodelay(true);
+                builder.build(tcp)
+            };
+            let dest: Multiaddr = DEST.parse().unwrap();
+            let behaviour = ping::Behaviour::new(ping::Config::new().with_keep_alive(true));
+            let mut swarm = Swarm::new(transport, behaviour, peer_id);
+            swarm.dial(dest).unwrap();
```
Comment (polarbeargo) [inline network/tests/connection_mainnet.rs:51]: Excellent work on testing the p2p network by transport config file 🚀 !

**Review by polarbeargo — COMMENTED** [review]
Overall it's awesome work on implementing network crate and integrating module p2p 🚀! Good start!

File: network/src/p2p/builder.rs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        let pnet_config = PnetConfig::new(shared_key);
+        let mplex_config = config.get_mplex_config();
+        let noise_keys = noise::Keypair::<noise::X25519Spec>::new()
+            .into_authentic(&keypair)
+            .expect("Signing libp2p-noise static DH keypair failed.");
+        Self {
+            pnet_config,
+            mplex_config,
+            noise_keys,
+            timeout: Duration::from_secs(60),
```
Comment (creativcoder) [inline network/src/p2p/builder.rs:54]: this also looks like can be moved to TransportConfig and made configurable.

File: network/src/p2p/builder.rs [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    noise::{self, AuthenticKeypair, X25519Spec},
+    pnet::PnetConfig,
+    PeerId, Transport,
+};
+use libp2p_mplex::MplexConfig;
+use std::{borrow::Borrow, time::Duration};
+
+/// Builds libp2p transport for mina with various configurations
+#[derive(Clone)]
+pub struct MinaTransportBuilder {
```
Comment (creativcoder) [inline network/src/p2p/builder.rs:18]: naming: can we remove the mina prefix and just have `TransportBuilder`? reason: the project name is likely to change. likewise for other entities wherever possible.

**Review by creativcoder — APPROVED** [review]
nothing major, just minor comments, looks good for merge

File: network/tests/p2p.rs [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+#[cfg(test)]
+pub mod tests {
+    use std::time::Duration;
+
+    use libp2p::{futures::StreamExt, ping, swarm::SwarmEvent, tcp::TcpConfig, Multiaddr, Swarm};
+    use mina_network::p2p::{MinaTransportBuilder, MinaTransportConfig, MAINNET_CONFIG};
+
+    const DEST: &str =
+        "/ip4/95.217.106.189/tcp/8302/p2p/12D3KooWSxxCtzRLfUzoxgRYW9fTKWPUujdvStuwCPSPUN3629mb";
```
Comment (noot) [inline network/tests/p2p.rs:12]: where is this from? is it a mina bootnode?

**Review by noot — APPROVED** [review]
wow, rust libp2p is much more complicated than go, haha

File: network/tests/p2p.rs (hunk 10 above)
Comment (hanabi1224) [reply network/tests/p2p.rs:12]: This is one of the seed nodes

File: network/src/p2p/builder.rs (hunk 8 above)
Comment (hanabi1224) [reply network/src/p2p/builder.rs:54]: There will be static immutable instances for different preset TransportConfig(s), but timeout should be mutable (e.g. via CLI args),

File: network/src/p2p/builder.rs [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    pub fn with_config(mut self, config: &MinaTransportConfig) -> Self {
+        let shared_key = config.get_shared_key();
+        self.pnet_config = PnetConfig::new(shared_key);
+        self.mplex_config = config.get_mplex_config();
+        self
+    }
+
+    pub fn with_mainnet_config(self) -> Self {
+        self.with_config(&MAINNET_CONFIG)
+    }
```
[the remark this answers is not in the document]
Comment (hanabi1224) [reply network/src/p2p/builder.rs:70]: See line 73 `pub fn with_timeout(mut self, timeout: Duration) -> Self`

File: network/src/p2p/builder.rs (hunk 9 above)
Comment (hanabi1224) [reply network/src/p2p/builder.rs]: Fixed

**Pull request merged**